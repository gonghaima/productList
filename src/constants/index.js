import keyMirror from 'fbjs/lib/keyMirror';

/**
 * @namespace Constants
 * @desc App constants
 */

/**
 * @constant {Object} ActionTypes
 * @memberof Constants
 */
export const ActionTypes = keyMirror({
  SWITCH_MENU: undefined,
  EXCEPTION: undefined,
  PRODUCT_GET_REPOS: undefined,
  PRODUCT_GET_REPOS_SUCCESS: undefined,
  PRODUCT_GET_REPOS_FAILURE: undefined,
  SELECTION_CHANGE: undefined,
  PAGE_CHANGE: undefined,
});

/**
 * @constant {Object} STATUS
 * @memberof Constants
 */
export const STATUS = {
  IDLE: 'idle',
  RUNNING: 'running',
  READY: 'ready',
  SUCCESS: 'success',
  ERROR: 'error',
};
