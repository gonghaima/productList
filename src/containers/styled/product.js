import styled from 'styled-components';
import { titleColor, borderColor, itemBgColor, subTitleColor, focusedColor } from 'modules/theme';

import { Box, utils } from 'styled-minimal';

const { responsive, spacer } = utils;

const Title = styled.h2`
  color: ${titleColor};
  padding: 0 10px;
  margin: 0;
`;

const Summary = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: space-between;
  padding: 0 10px;
  color: ${subTitleColor};
`;
const SummaryItem = styled.div`
  margin: 0;
  width: 10%;
  select {
    background-color: #f6f6f6 !important;
  }
`;

const ProductCount = styled.span`
  vertical-align: sub;
`;

const Divider = styled.hr`
  display: block;
  height: 1px;
  border: 0;
  border-top: 1px solid ${borderColor};
  margin: 0.7em 0.4em;
  padding: 0;
`;
const ProductGrid = styled.ul`
  display: grid;
  grid-auto-flow: row;
  grid-gap: ${spacer(2)};
  grid-template-columns: 100%;
  list-style: none;
  margin: 0px ${spacer(4)} auto 0;
  padding: 0;
  /* stylelint-disable */
  ${/* istanbul ignore next */ p =>
    responsive({
      ix: `
        grid-gap: ${spacer(3)(p)};
        width: 90%;
      `,
      md: `
        grid-template-columns: repeat(2, 1fr);
        width: 100%;
      `,
      lg: `
        grid-template-columns: repeat(3, 1fr);
      `,
      xl: `
        grid-gap: ${spacer(4)(p)};
        grid-template-columns: repeat(4, 1fr);
      `,
    })};
  /* stylelint-enable */

  > li {
    display: flex;
  }
`;

const Item = styled(Box)`
  background: ${itemBgColor};
  align-items: center;
  border: solid 0.1rem ${borderColor};
  border-radius: 0.4rem;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  overflow: hidden;
  padding: ${spacer(3)};
  text-align: center;
  width: 100%;
  /* stylelint-disable */
  ${/* istanbul ignore next */ p =>
    responsive({
      md: `
        padding: ${spacer(3)(p)} ${spacer(0)(p)};
      `,
      lg: `
        padding: ${spacer(2)(p)} ${spacer(0)(p)};
      `,
    })};
  /* stylelint-enable */

  > a {
    margin-bottom: ${spacer(2)};
  }

  img {
    height: 20rem;
  }
`;

const ImageSection = styled.div`
  padding: 6px;
  margin: 0;
`;
const DetailsSection = styled.div`
  border-top: 1px solid ${borderColor};
  text-align: left;
  padding-left: ${spacer(3)};
  width: 100%;
  display: block;
  margin-bottom: ${spacer(3)};

  a {
    display: block;
  }
`;

const ItemTitle = styled.h4`
  color: ${titleColor};
  margin: ${spacer(3)} ${spacer(0)} ${spacer(1)} ${spacer(0)};
`;

const ItemDescription = styled.h4`
  color: ${subTitleColor};
  margin: 0px 0px ${spacer(2)} 0px;
`;

const ItemPrice = styled.h6`
  color: ${focusedColor};
  margin: 0;
`;

const Pagination = styled.div`
  display: flex;
  flex-direction: row-reverse;
  ul {
    display: inline-block;
    padding-left: 15px;
    padding-right: 15px;
  }

  li {
    display: inline-block;
  }
  a {
    z-index: 3;
    color: #000;
    cursor: pointer;
    position: relative;
    float: left;
    padding: 6px 12px;
    margin-left: -1px;
    line-height: 1.42857143;
    text-decoration: none;
  }
  .active a:hover,
  .active a {
    background: #fff;
    outline: unset;
    border-bottom: 3px solid ${focusedColor};
  }
  .disabled a {
    color: ${subTitleColor};
    cursor: not-allowed;
  }
`;

export {
  Title,
  Summary,
  SummaryItem,
  ProductCount,
  Divider,
  ProductGrid,
  Item,
  ImageSection,
  DetailsSection,
  ItemTitle,
  ItemDescription,
  ItemPrice,
  Pagination,
};
