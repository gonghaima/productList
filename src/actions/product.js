// @flow
/**
 * @module Actions/User
 * @desc User Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const { productGetRepos: getRepos } = createActions({
  [ActionTypes.PRODUCT_GET_REPOS]: query => ({ query }),
});
