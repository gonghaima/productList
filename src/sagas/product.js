/**
 * @module Sagas/Product
 * @desc Product
 */

import { all, call, put, takeLatest } from 'redux-saga/effects';
import { request } from 'modules/client';
import { ActionTypes } from 'constants/index';

/**
 * Get Repos
 *
 * @param {Object} action
 *
 */
export function* getProducts() {
  try {
    const response = yield call(request, '/api/product');
    yield put({
      type: ActionTypes.PRODUCT_GET_SUCCESS,
      payload: { data: (response && response.data) || [] },
    });
  } catch (err) {
    yield put({
      type: ActionTypes.PRODUCT_GET_REPOS_FAILURE,
      payload: err,
    });
  }
}

/**
 * Product Sagas
 */
export default function* root() {
  yield all([takeLatest(ActionTypes.PRODUCT_GET_MAKE, getProducts)]);
}
