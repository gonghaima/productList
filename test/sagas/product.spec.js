import { expectSaga } from 'redux-saga-test-plan';
import product, { getProducts } from 'sagas/product';
import { ActionTypes } from 'constants/index';

jest.mock('modules/client', () => ({
  request: () => ({ items: [] }),
}));

describe('product', () => {
  it('should have the expected watchers', done =>
    expectSaga(product)
      .run({ silenceTimeout: true })
      .then(saga => {
        expect(saga).toMatchSnapshot();
        done();
      }));

  it('should have the product saga', () =>
    expectSaga(getProducts, { payload: { query: 'react' } })
      .put({
        type: ActionTypes.PRODUCT_GET_SUCCESS,
        payload: {
          data: [],
        },
      })
      .run());
});
