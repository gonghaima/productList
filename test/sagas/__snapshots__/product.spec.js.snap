// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product should have the expected watchers 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "PRODUCT_GET_REPOS",
          "@@redux-saga-test-plan/json/function/getRepos",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "PRODUCT_GET_REPOS",
      },
    },
  ],
}
`;
